README.md
![alt tag](https://raw.githubusercontent.com/lateralblast/modest/master/modest.gif)

MODEST
======

Multi OS Deployment Engine Server Tool

Introduction
------------

A Ruby script to configure server and client configuration for PXE/DHCP and
BOOTP based install services, eg:

- Solaris 11 Automated Installer (AI) service
- Solaris 10 (and earlier) Jumpstart services
- Centos, Scientific, OEL, Fedora, and RHEL Kickstart services
- Ubuntu Linux Preseed services
- SuSE Enterprise Linux AutoYast services
- ESXi Kickstart services
- Simplify creation, deletion, import and export of VirtualBox VMs
- Simplify creation, deletion, import and export of VMware Fusion VMs
- Simpllfy creation of Solaris Zones
- Simplify creation of Solaris VM Server for SPARC Control Domains
- Simplify creation of Solaris VM Server for SPARC Guest Domains
- Linux Containers (Currently Ubuntu support only)
- OpenBSD PXE boot service
- Initial CoreOS PXE boot support
- Provide completely headless VM support

This script is a wrapper which sits on top of the exisitng tools. It changes
defaults to be more suitable to customer environments, and sanity checks server
and client configuration to ensure the chance of human error is reduced.

By doing this the script reduces the time lost debugging installations,
which in complex environments can be time consuming. It also reduces the time
to deploy servers and increases the consistency of installations.

I've added completely headless VM support via virtual serial port so that
installs via serial over IPMI can also be tested for environments that do not
have remove KVM capability. For example ESXi can be installed into a VM in
completely headless mode, completely simulating driving the install via serial.

For an example of a completely serial install take a look at the [Video Examples](https://github.com/lateralblast/modest/wiki/7.-Video-Examples)
section.

This script will also provide the ability to install Solaris 2.6-10, Solaris 11,
ESXi, and Linux (SuSE, Ubuntu, RedHat and Centos) all from one server and OS,
consolidating and simplifying installation services. This can be hosted on a
physical server or a VM.

In particular it can be used with a laptop to provide installation services via
a cross-over cable. This is expecially useful for resolving issues with servers
and installing firmware. It can be used in combination with the firmware and
patch repository script (goofball) to install patches, firmware and SRUs.

License
-------

This software is licensed as CC-BA (Creative Commons By Attrbution)

http://creativecommons.org/licenses/by/4.0/legalcode

Usage
-----

Modest is designed to be used in a number of scenarios:

- Testing vendor OS deployment technologies (e.g. Kickstart, Jumpstart etc)
  - PXE boot
  - Pre and post install scripts
- A environment for quickly deploying VMs for testing purposes
  - In a similar fashion to veewee or vagrant
- An environment for quickly deploying servers in a Datacenter or testlab via cross-over cable

A good place to get started is the [Usage](https://github.com/lateralblast/modest/wiki/3.-Usage)
Wiki page which includes a section on getting started.

Wiki Index
----------

[Background](https://github.com/lateralblast/modest/wiki/1.-Background)

[Features](https://github.com/lateralblast/modest/wiki/2.-Features)

[Usage](https://github.com/lateralblast/modest/wiki/3.-Usage)

- [VirtualBox](https://github.com/lateralblast/modest/wiki/3.1.-VirtualBox-Examples)
- [VMware Fusion](https://github.com/lateralblast/modest/wiki/3.2.-VMware-Fusion-Examples)

[Server](https://github.com/lateralblast/modest/wiki/4.-Server)

- [AI](https://github.com/lateralblast/modest/wiki/4.1.-AI-Server-Examples)
- [Jumpstart](https://github.com/lateralblast/modest/wiki/4.2.-Jumpstart-Server-Examples)
- [Kickstart](https://github.com/lateralblast/modest/wiki/4.3.-Kickstart-Server-Examples)
- [LDom](https://github.com/lateralblast/modest/wiki/4.4.-LDom-Server-Examples)
- [LXC](https://github.com/lateralblast/modest/wiki/4.5.-Linux-Container-Server-Examples)
- [Preseed](https://github.com/lateralblast/modest/wiki/4.6.-Preseed-Server-Examples)
- [vShpere](https://github.com/lateralblast/modest/wiki/4.7.-Sphere-Server-Examples)
- [Zones](https://github.com/lateralblast/modest/wiki/4.8.-Zone-Server-Examples)

[Client](https://github.com/lateralblast/modest/wiki/5.-Client)

- [AI](https://github.com/lateralblast/modest/wiki/5.1.-AI-Client-Example)
- [Jumpstart](https://github.com/lateralblast/modest/wiki/5.2.-Jumpstart-Client-Examples)
- [Kickstart](https://github.com/lateralblast/modest/wiki/5.3.-Kickstart-Client-Examples)
- [LDom](https://github.com/lateralblast/modest/wiki/5.4.-LDom-Client-Examples)
- [LXC](https://github.com/lateralblast/modest/wiki/5.5.-Linux-Container-Client-Examples)
- [Preseed](https://github.com/lateralblast/modest/wiki/5.6.-Preseed-Client-Examples)
- [vSphere](https://github.com/lateralblast/modest/wiki/5.7.-vSphere-Client_Examples)
- [Zones](https://github.com/lateralblast/modest/wiki/5.8.-Zone-Client-Examples)

[Video Examples](https://github.com/lateralblast/modest/wiki/7.-Video-Examples)

- [ESXi](https://github.com/lateralblast/modest/wiki/7.1.-ESXi-Video-Examples)

[Other Examples](https://github.com/lateralblast/modest/wiki/8.-Other-Examples)

- [Deploying VCSA](https://github.com/lateralblast/modest/wiki/8.1.-Deploying-VCSA)

Features
--------

Linux Container Creation

- Installs required packages and sets up network for public facing containers
  that can be connected to via ssh (Currently Ubuntu support only)
- Creates containers and sets up network (Currently Ubuntu support only)

Oracle VM Server for SPARC

- Setup Control Domain based on a set of questions
- Cleate Guest Domains based on a set of questions

VirtualBox and VMware Fusion VM creation

- Can create VMs to speed up the automation of testing new images
- Can run VMs headless and connect via serial allowing remote installation

Solaris Zone Creation

- Asks a set of questions to generate a template and then install a zone

Solaris 11 AI

- If a local repository exists it is used to configure installation services
- If a local repository is not present it mounts the repository ISO and
  configures it
- Changes the default install service manifest to reboot servers after
  installation
- Changes the default install service to use the local repository rather than
  the default Oracle one
- Changes the default grub configuration default to an autamated install rather
  than an interactive one
- Disables the default range based DHCP configuration and adds clients on an
  individual basis
  - Reduces the chances of conflicts with existing DHCP services, and
  - Resolves the issue of DHCP licenses running out
- Creates an individual configuration profile for each client
  - Configures network and other services

changelog
# Changes
# 0.0.1: Tue 17 Dec 2013 12:25:55 EST
# Fixed Dir.* usage to be compatible on ruby 1.8
# 0.0.2: Tue 17 Dec 2013 12:33:32 EST
# Fixed AI publisher typo
# 0.0.3: Tue 17 Dec 2013 13:36:13 EST
# Fixed system command output
# 0.0.4: Tue 17 Dec 2013 16:44:33 EST
# Fixed processing of questions
# 0.0.5: Wed 18 Dec 2013 07:10:32 EST
# Added code to delete AI client and service
# 0.0.6: Wed 18 Dec 2013 07:50:35 EST
# Added code to be able to create only one AI architecture service
# 0.0.7: Wed 18 Dec 2013 14:27:02 EST
# Added AI service manifest creation code
# 0.0.8: Wed 18 Dec 2013 18:42:33 EST
# Improved verbose output
# 0.0.9: Wed 18 Dec 2013 19:06:36 EST
# Fixed ISO file determination
# 0.1.0: Wed 18 Dec 2013 22:15:55 EST
# Added to delete service
# 0.1.1: Thu 19 Dec 2013 10:20:38 EST
# Cleaned up server configuration code
# 0.1.2: Thu 19 Dec 2013 10:51:20 EST
# Fixed service creation
# 0.1.3: Thu 19 Dec 2013 12:20:31 EST
# Fixed auto_install directory to include release information
# 0.1.4: Thu 19 Dec 2013 14:37:25 EST
# Fixed issue with XML
# 0.1.5: Thu 19 Dec 2013 15:57:03 EST
# Fixed order of questions
# 0.1.6: Thu 19 Dec 2013 16:21:00 EST
# Fixed pkg url in manifest
# 0.1.7: Thu 19 Dec 2013 20:04:44 EST
# Cleaned up deletion code
# 0.1.8: Thu 19 Dec 2013 22:26:47 EST
# Fixed bugs with AI manifest creation
# 0.1.9: Fri 20 Dec 2013 19:30:28 EST
# Added additional questions to profile
# 0.2.0: Fri 20 Dec 2013 20:03:01 EST
# Fixed bugs with error handling and XML creation
# 0.2.1: Fri 20 Dec 2013 21:10:40 EST
# Added code to change timeout and default menu item in netboot grub.cfg
# 0.2.2: Fri 20 Dec 2013 21:26:36 EST
# Fixed XML output bugs
# 0.2.3: Sun 22 Dec 2013 15:40:34 EST
# Added code to use local file based repository if it exists
# 0.2.4: Sun 22 Dec 2013 16:25:34 EST
# Added code to extract release version from repository
# 0.2.5: Mon 23 Dec 2013 18:15:04 EST
# Added code to remove range entry from dhcpd config
# 0.2.6: Mon 23 Dec 2013 19:14:37 EST
# Added code to restart dhcp service
# 0.2.7: Mon 23 Dec 2013 19:24:20 EST
# Fixed bug with backing up dhcp config
# 0.2.8: Tue 24 Dec 2013 04:01:06 EST
# Fixed bug with editing grub.cfg
# 0.2.9: Thu 26 Dec 2013 14:33:14 EST
# Added code for alternate package repository to deploy puppet packages
# 0.3.0: Thu 26 Dec 2013 15:06:17 EST
# Added more value checking
# 0.3.1: Thu 26 Dec 2013 15:08:48 EST
# Updated usage information
# 0.3.2: Thu 26 Dec 2013 18:11:35 EST
# Fixed alternate repository creation and deletion
# 0.3.3: Thu 26 Dec 2013 18:59:44 EST
# Fixed alternate repository prefix
# 0.3.4: Thu 26 Dec 2013 19:39:51 EST
# Fixed package repository check
# 0.3.5: Thu 26 Dec 2013 20:37:30 EST
# Added code to install and uninstall packages
# 0.3.6: Fri 27 Dec 2013 08:59:43 EST
# Fixed package creation
# 0.3.7: Fri 27 Dec 2013 10:09:58 EST
# Fixed package version check
# 0.3.8: Fri 27 Dec 2013 12:55:37 EST
# Fixed package dependency handling
# 0.3.9: Fri 27 Dec 2013 14:49:38 EST
# Fixed grub.cfg update code
# 0.4.0: Fri 27 Dec 2013 15:37:09 EST
# Added ruby to base installation
# 0.4.1: Sat 28 Dec 2013 13:54:14 EST
# Fixed multiline output
# 0.4.2: Sat 28 Dec 2013 15:37:35 EST
# Added Linux ISO mount and copy code
# 0.4.3: Sat 28 Dec 2013 17:59:49 EST
# Added code to add and remove apache alias for kickstart directory
# 0.4.4: Sat 28 Dec 2013 19:42:06 EST
# Added code to list AI services

methods/in_osx.rb
# Common OS X routines

# Get OSX default rotute interface

def get_osx_gw_if_name()
  message    = "Getting:\tInterface name of default router"
  command    = "netstat -rn |grep ^default |head -1 |awk '{print $6}'"
  gw_if_name = execute_command(message,command)
  gw_if_name = gw_if_name.chomp
  return gw_if_name
end

# Check PF is configure on OS X 10.10

def check_osx_pfctl(gw_if_name,if_name)
  pf_file = $work_dir+"/pfctl_config"
  if File.exist?(pf_file)
    File.delete(pf_file)
  end
  output = File.open(pf_file,"w")
  if $verbose_mode == 1
    puts "Enabling forwarding between "+gw_if_name+" and "+if_name
  end
  output.write("nat on #{gw_if_name} from #{if_name}:network to any -> (#{gw_if_name})\n")
  output.write("pass inet proto icmp all\n")
  output.write("pass in on #{if_name} proto udp from any to any port domain keep state\n")
  output.write("pass in on #{if_name} proto tcp from any to any port domain keep state\n")
  output.write("pass quick on #{gw_if_name} proto udp from any to any port domain keep state\n")
  output.write("pass quick on #{gw_if_name} proto tcp from any to any port domain keep state\n")
  output.close
  message = "Enabling:\tPacket filtering"
  command = "sudo pfctl -e"
  execute_command(message,command)
  message = "Loading:\yFilters from "+pf_file
  command = "sudo pfctl -F all -f #{pf_file}"
  execute_command(message,command)
  return
end

# check NATd is running and configured on OS X 10.9 and earlier
# Useful info on pfctl here http://patrik-en.blogspot.com.au/2009/10/nat-in-virtualbox-with-freebsd-and-pf.html

def check_osx_nat(gw_if_name,if_name)
  message = "Checking:\tIP forwarding is enabled"
  command = "sudo sysctl -a net.inet.ip.forwarding |awk '{print $2}'"
  output  = execute_command(message,command)
  output  = output.chomp
  if output =~ /0/
    message = "Enabling:\tIP forwarding"
    command = "sudo sysctl net.inet.ip.forwarding=1"
    execute_command(message,command)
  end
  message = "Checking:\tRule for IP forwarding has been created"
  if $os_rel.match(/^14/)
    command = "sudo pfctl -a '*' -sr 2>&1 |grep 'pass quick on #{gw_if_name}'"
  else
    command = "sudo ipfw list |grep 'any to any via #{gw_if_name}'"
  end
  output  = execute_command(message,command)
  if !output.match(/#{gw_if_name}/)
    message = "Enabling:\tNATd to forward traffic on "+gw_if_name
    if $os_rel.match(/^14/)
      check_osx_pfctl(gw_if_name,if_name)
    else
      command "sudo ipfw add 100 divert natd ip from any to any via #{gw_if_name}"
      execute_command(message,command)
    end
  end
  if !$os_rel.match(/^14/)
    message = "Checking:\tNATd is running"
    command = "ps -ef |grep '#{gw_if_name}' |grep natd |grep 'same_ports'"
    output  = execute_command(message,command)
    if !output.match(/natd/)
      message = "Starting:\tNATd to foward packets between "+if_name+" and "+gw_if_name
      command = "sudo /usr/sbin/natd -interface #{gw_if_name} -use_sockets -same_ports -unregistered_only -dynamic -clamp_mss -enable_natportmap -natportmap_interface #{if_name}"
      execute_command(message,command)
    end
  end
  return
end

# Tune OS X NFS

def tune_osx_nfs()
  nfs_file   = "/etc/nfs.conf"
  nfs_params = ["nfs.server.nfsd_threads = 64","nfs.server.reqcache_size = 1024","nfs.server.tcp = 1","nfs.server.udp = 0","nfs.server.fsevents = 0"]
  nfs_params.each do |nfs_tune|
    nfs_tune = "nfs.client.nfsiod_thread_max = 64"
    message  = "Checking:\tNFS tuning"
    command  = "cat #{nfs_file} |grep '#{nfs_tune}'"
    output   = execute_command(message,command)
    if !output.match(/#{nfs_tune}/)
      backup_file(nfs_file)
      message = "Tuning:\tNFS"
      command = "echo '#{nfs_tune}' >> #{nfs_file}"
      execute_command(message,command)
    end
  end
  return
end

# Get Mac disk name

def get_osx_disk_name()
  message = "Getting:\tRoot disk device ID"
  command = "df |grep '/$' |awk '{print \\$1}'"
  output  = execute_command(message,command)
  disk_id = output.chomp
  message = "Getting:\tVolume name for "+disk_id
  command = "diskutil info #{disk_id} | grep 'Volume Name' |cut -f2 -d':'"
  output  = execute_command(message,command)
  volume  = output.chomp.gsub(/^\s+/,"")
  return volume
end

# Check OSX Puppet install

def check_osx_puppet_install()
  pkg_list = {}
  use_rvm  = 0
  pkg_list["facter"] = $facter_version
  pkg_list["hiera"]  = $hiera_version
  pkg_list["puppet"] = $puppet_version
  base_url  = "http://downloads.puppetlabs.com/mac/"
  local_dir = $work_dir+"/dmg"
  check_dir_exists(local_dir)
  pkg_list.each do |key, value|
    test_file = "/usr/bin/"+key
    if !File.exist?(test_file)
      file_name   = key+"-"+value
      dmg_name    = file_name+".dmg"
      local_pkg   = key+"-"+value+".pkg"
      remote_file = base_url+"/"+dmg_name
      local_file  = local_dir+"/"+dmg_name
      if !File.exist?(local_file)
        wget_file(remote_file,local_file)
      end
      message = "Mounting:\tDisk image "+local_file
      command = "hdiutil mount #{local_file}"
      execute_command(message,command)
      local_pkg = "/Volumes/"+file_name+"/"+local_pkg
      volume    = get_osx_disk_name()
      volume    = "/Volumes/"+volume
      message   = "Installing:\tPackage "+local_pkg
      command   = "installer -package #{local_pkg} -target '#{volume}'"
      execute_command(message,command)
      if key.match(/puppet/)
        message = "Checking:\tRuby version"
        command = "which ruby"
        output  = execute_command(message,command)
        if output.match(/rvm/)
          use_rvm  = 1
          message  = "Storing:\tRVM Ruby version"
          command  = "rvm current"
          output   = execute_command(message,command)
          rvm_ruby = output.chomp
          message  = "Setting:\tRVM to use system ruby"
          command  = "rvm use system"
          execute_command(message,command)
        end
        message = "Creating:\tPuppet group"
        command = "puppet resource group puppet ensure=present"
        execute_command(message,command)
        message = "Creating:\tPuppet user"
        command = "puppet resource user puppet ensure=present gid=puppet shell='/sbin/nologin'"
        execute_command(message,command)
        etc_dir = "/etc/puppet"
        check_dir_exists(etc_dir)
        message = "Creating:\tPuppet directory"
        command = "mkdir -p /var/lib/puppet ; mkdir -p /etc/puppet/manifests ; mkdir -p /etc/puppet/ssl"
        execute_command(message,command)
        message = "Fixing:\tPuppet permissions"
        command = "chown -R puppet:puppet  /var/lib/puppet ; chown -R puppet:puppet  /etc/puppet"
        execute_command(message,command)
        if use_rvm == 1
          message = "Reverting:\tRVM to use "+rvm_ruby
          command = "rvm use rvm_ruby"
          execute_command(message,command)
        end
      end
      local_vol = "/Volumes/"+key+"-"+value
      message   = "Unmounting:\t"+local_vol
      command   = "umount "+local_vol
      execute_command(message,command)
    end
  end

  return
end

# Create OS X Puppet agent plist file

def create_osx_puppet_agent_plist()
  xml_output = []
  plist_file = "/Library/LaunchDaemons/com.puppetlabs.puppet.plist"
  tmp_file   = "/tmp/puppet.plist"
  plist_name = "com.puppetlabs.puppet"
  puppet_bin = "/usr/bin/puppet"
  message    = "Checking:\tPuppet configruation"
  command    = "cat #{plist_file} | grep 'agent'"
  output     = execute_command(message,command)
  if !output.match(/#{$default_net}/)
    xml = Builder::XmlMarkup.new(:target => xml_output, :indent => 2)
    xml.instruct! :xml, :version => "1.0", :encoding => "UTF-8"
    xml.declare! :DOCTYPE, :plist, :PUBLIC, :'"-//Apple Computer//DTD PLIST 1.0//EN"', :'"http://www.apple.com/DTDs/PropertyList-1.0.dtd"'
    xml.plist(:version => "1.0") {
      xml.dict {
        xml.key("EnvironmentVariables")
        xml.dict {
          xml.key("PATH")
          xml.string("/sbin:/usr/sbin:/bin:/usr/bin")
          xml.key("RUBYLIB")
          xml.string("/usr/lib/ruby/site_ruby/1.8/")
        }
        xml.key("label")
        xml.string(plist_name)
        xml.key("OnDemand") ; xml.false
        xml.key("ProgramArguments")
        xml.array {
          xml.string(puppet_bin)
          xml.string("agent")
          xml.string("--verbose")
          xml.string("--no-daemonize")
          xml.string("--log-dest")
          xml.string("console")
        }
      }
      xml.key("RunAtLoad") ; xml.true
      xml.key("ServiceIPC") ; xml.false
      xml.key("StandardErrorPath")
      xml.string("/var/log/puppet/puppet.err")
      xml.key("StandardOutPath")
      xml.string("/var/log/puppet/puppet.out")
    }
    file=File.open(tmp_file,"w")
    xml_output.each do |item|
      file.write(item)
    end
    file.close
    message = "Creating:\tService file "+plist_file
    command = "cp #{tmp_file} #{plist_file} ; rm #{tmp_file} ; chown root:wheel #{plist_file} ; chmod 644 #{plist_file}"
    execute_command(message,command)
  end
  return
end

# Create OS X Puppet master plist file

def create_osx_puppet_master_plist()
  xml_output = []
  plist_file = "/Library/LaunchDaemons/com.puppetlabs.puppetmaster.plist"
  tmp_file   = "/tmp/puppetmaster.plist"
  plist_name = "com.puppetlabs.puppetmaster"
  puppet_bin = "/usr/bin/puppet"
  message    = "Checking:\tPuppet configruation"
  command    = "cat #{plist_file} | grep 'master'"
  output     = execute_command(message,command)
  if !output.match(/#{$default_net}/)
    xml = Builder::XmlMarkup.new(:target => xml_output, :indent => 2)
    xml.instruct! :xml, :version => "1.0", :encoding => "UTF-8"
    xml.declare! :DOCTYPE, :plist, :PUBLIC, :'"-//Apple Computer//DTD PLIST 1.0//EN"', :'"http://www.apple.com/DTDs/PropertyList-1.0.dtd"'
    xml.plist(:version => "1.0") {
      xml.dict {
        xml.key("EnvironmentVariables")
        xml.dict {
          xml.key("PATH")
          xml.string("/sbin:/usr/sbin:/bin:/usr/bin")
          xml.key("RUBYLIB")
          xml.string("/usr/lib/ruby/site_ruby/1.8/")
        }
        xml.key("label")
        xml.string(plist_name)
        xml.key("ProgramArguments")
        xml.array {
          xml.string(puppet_bin)
          xml.string("master")
          xml.string("--verbose")
          xml.string("--no-daemonize")
        }
      }
      xml.key("RunAtLoad") ; xml.true
      xml.key("ServiceIPC") ; xml.false
      xml.key("StandardErrorPath")
      xml.string("/var/log/puppet/puppetmaster.err")
      xml.key("StandardOutPath")
      xml.string("/var/log/puppet/puppetmaster.out")
    }
    file=File.open(tmp_file,"w")
    xml_output.each do |item|
      file.write(item)
    end
    file.close
    message = "Creating:\tService file "+plist_file
    command = "cp #{tmp_file} #{plist_file} ; rm #{tmp_file} ; chown root:wheel #{plist_file} ; chmod 644 #{plist_file}"
    execute_command(message,command)
  end
  return
end

# Create OS X Puppet Agent plist

def create_osx_puppet_plist()
  return
end

# Create OS X Puppet Master plist

def create_osx_puppet_master_plist()
  return
end

# Check OX X Puppet plist

def check_osx_puppet_plist()
  plist_file = "/Library/LaunchDaemons/com.puppetlabs.puppet.plist"
  plist_name = "com.puppetlabs.puppet"
  if !File.exist?(plist_file)
    create_osx_puppet_plist()
    message = "Loading:\tPuppet Agent plist file "+plist_file
    command = "launchctl load -w #{plist_file}"
    execute_command(message,command)
    message = "Loading:\tStarting Puppet Agent "+plist_name
    command = "launchctl start #{plist_name}"
    execute_command(message,command)
  end
  plist_file = "/Library/LaunchDaemons/com.puppetlabs.puppetmaster.plist"
  plist_name = "com.puppetlabs.puppetmaster"
  if !File.exist?(plist_file)
    create_osx_puppet_master_plist()
    message = "Loading:\tPuppet Master plist file "+plist_file
    command = "launchctl load -w #{plist_file}"
    execute_command(message,command)
    message = "Loading:\tStarting Puppet Master "+plist_name
    command = "launchctl start #{plist_name}"
    execute_command(message,command)
  end
  return
end

# Check OS X apache config

def check_osx_apache()
  ssl_dir = "/private/etc/apache2/ssl"
  check_dir_exists(ssl_dir)
  server_key = ssl_dir+"/server.key"
  if !File.exist?(server_key)
    message = "Generating:\tApache SSL Server Key "+server_key
    command = "ssh-keygen -f #{server_key}"
    execute_command(message,command)
  end
  return
end

# Create OS X Puppet config

def create_osx_puppet_config()
  tmp_file    = "/tmp/puppet_config"
  puppet_file = "/etc/puppet/puppet.conf"
  if !File.exist?(puppet_file)
    config = []
    config.push("[main]")
    config.push("pluginsync = true")
    config.push("server = #{$default_host}")
    config.push("")
    config.push("[master]")
    config.push("vardir = /var/lib/puppet")
    config.push("libdir = $vardir/lib")
    config.push("ssldir = /etc/puppet/ssl")
    config.push("certname = #{$default_host}")
    config.push("")
    config.push("[agent]")
    config.push("vardir = /var/lib/puppet")
    config.push("libdir = $vardir/lib")
    config.push("ssldir = /etc/puppet/ssl")
    config.push("certname = #{$default_host}")
    config.push("")
    file = File.open(tmp_file,"w")
    config.each do |line|
      output = line+"\n"
      file.write(output)
    end
    file.close
    message = "Creating:\tPuppet configuration file "+puppet_file
    command = "cp #{tmp_file} #{puppet_file} ; rm #{tmp_file}"
    execute_command(message,command)
    print_contents_of_file(puppet_file)
  end
  return
end

# Create OS X Puppet config

def create_osx_puppet_config()
  return
end

# Check OS X Puppet

def check_osx_puppet()
  check_osx_puppet_install()
  check_osx_puppet_plist()
  create_osx_puppet_config()
  return
end

# Check OS X IPS

def check_osx_ips()
  python_bin = "/usr/bin/python"
  pip_bin    = "/usr/bin/pip"
  setup_url  = "https://bitbucket.org/pypa/setuptools/raw/bootstrap/ez_setup.py"
  if !File.symlink?(pip_bin)
    message = "Installing:\tPip"
    command = "/usr/bin/easy_install --prefix=/usr pip"
    execute_command(message,command)
    message = "Updating:\tSetuptools"
    command = "wget #{setup_url} -O |sudo #{python_bin}"
    execute_command(message,command)
    ["simplejson","coverage","pyOpenSSL","mercurial"].each do |module_name|
      message = "Installing:\tPython module "+module_name
      command = "#{pip_bin} install #{module_name}"
      execute_command(message,command)
    end
  end
  python_ver = %x[#{python_bin} --version |awk '{print $2}']
  python_ver = python_ver.chomp.split(/\./)[0..1].join(".")
  module_dir = "/usr/local/lin/python"+python_ver+"/site-packages"
  pkg_dest_dir = module_dir+"/pkg"
  check_dir_exists(pkg_dest_dir)
  hg_bin = "/usr/local/bin/hg"
  if !File.exist?(hg_bin)
    message = "Installing:\tMercurial"
    command = "brew install mercurial"
    execute_command(message,command)
  end
  pkgrepo_bin = "/usr/local/bin/pkgrepo"
  if !File.exist?(pkgrepo_bin)
    ips_url = "https://hg.java.net/hg/ips~pkg-gate"
    message = "Downloading:\tIPS source code"
    command = "cd #{$work_dir} ; hg clone #{ips_url} ips"
    execute_command(message,command)
  end
  return
end

# Check OSX service is enabled

def check_osx_service_is_enabled(service)
  service     = get_service_name(service)
  plist_file  = "/Library/LaunchDaemons/"+service+".plist"
  if !File.exist?(plist_file)
    plist_file = "/System"+plist_file
  end
  if !File.exist?(plist_file)
    puts "Warning:\tLaunch Agent not found for "+service
    exit
  end
  tmp_file  = "/tmp/tmp.plist"
  message   = "Checking:\tService "+service+" is enabled"
  if service.match(/dhcp/)
    command   = "cat #{plist_file} | grep Disabled |grep true"
  else
    command   = "cat #{plist_file} | grep -C1 Disabled |grep true"
  end
  output    = execute_command(message,command)
  if !output.match(/true/)
    if $verbose_mode == 1
      puts "Information:\t"+service+" enabled"
    end
  else
    backup_file(plist_file)
    copy      = []
    check     = 0
    file_info = IO.readlines(plist_file)
    file_info.each do |line|
      if line.match(/Disabled/)
        check = 1
      end
      if line.match(/Label/)
        check = 0
      end
      if check == 1 and line.match(/true/)
        copy.push(line.gsub(/true/,"false"))
      else
        copy.push(line)
      end
    end
    File.open(tmp_file,"w") {|file| file.puts copy}
    message = "Enabling:\t"+service
    command = "cp #{tmp_file} #{plist_file} ; rm #{tmp_file}"
    execute_command(message,command)
    message = "Loading:\t"+service+" profile"
    command = "launchctl load -w #{plist_file}"
    execute_command(message,command)
  end
  return
end

# Check TFTPd enabled on OS X

def check_osx_tftpd()
  service = "tftp"
  check_osx_service_is_enabled(service)
  return
end

# Check OS X brew package

def check_brew_pkg(pkg_name)
  message = "Checking:\tBrew package "+pkg_name
  command = "brew info #{pkg_name}"
  output  = execute_command(message,command)
  return output
end

# Install package with brew

def install_brew_pkg(pkg_name)
  pkg_status = check_brew_pkg(pkg_name)
  if !pkg_status.match(/[0-9]/)
    message = "Installing:\tPackage "+pkg_name
    command = "brew install #{pkg_name}"
    execute_command(message,command)
  end
  return
end

# Check OS X dnsmasq (used for puppet)

def check_osx_dnsmasq()
  pkg_name = "dnsmasq"
  install_brew_pkg(pkg_name)
  plist_file   = "/Library/LaunchDaemons/homebrew.mxcl.dnsmasq.plist"
  dnsmasq_file = "/usr/local/etc/dnsmasq.conf"
  if !File.exist?(plist_file)
    message = "Creating:\tPlist file "+plist_file+" for "+pkg_name
    command = "cp -fv /usr/local/opt/dnsmasq/*.plist /Library/LaunchDaemons"
    execute_command(message,command)
    message = "Creating:\tConfiguration file "+plist_file
    command = "cp #{dnsmasq_file}.example #{dnsmasq_file}"
    execute_command(message,command)
    message = "Loaing:\tConfiguration for "+pkg_name
    command = "launchctl load -w #{plist_file}"
    execute_command(message,command)
  end
  return
end

# Check OSC DHCP installation on OS X

def check_osx_dhcpd_installed()
  brew_file   = "/usr/local/Library/Formula/isc-dhcp.rb"
  backup_file = brew_file+".orig"
  dhcpd_bin   = "/usr/local/sbin/dhcpd"
  if !File.symlink?(dhcpd_bin)
    pkg_name = "bind"
    install_brew_pkg(pkg_name)
    message  = "Updating:\tBrew sources list"
    command  = "brew update"
    execute_command(message,command)
    message  = "Checking:\rOS X Version"
    command  = "sw_vers |grep ProductVersion |awk '{print $2}'"
    output   = execute_command(message,command)
    if output.match(/10\.9/)
      if File.exist?(brew_file)
        message = "Checking:\tVersion of ISC DHCPd"
        command = "cat #{brew_file} | grep url"
        output  = execute_command(message,command)
        if output.match(/4\.2\.5\-P1/)
          message = "Archiving:\tBrew file "+brew_file+" to "+backup_file
          command = "cp #{brew_file} #{backup_file}"
          execute_command(message,command)
          message = "Fixing:\tBrew configuration file "+brew_file
          command = "cat #{backup_file} | grep -v sha1 | sed 's/4\.2\.5\-P1/4\.3\.0rc1/g' > #{brew_file}"
          execute_command(message,command)
        end
        pkg_name = "isc-dhcp"
        install_brew_pkg(pkg_name)
      end
        message = "Creating:\tLaunchd service for ISC DHCPd"
        command = "cp -fv /usr/local/opt/isc-dhcp/*.plist /Library/LaunchDaemons"
        execute_command(message,command)
    end
    if !File.exist?($dhcpd_file)
      message = "Creating:\tDHCPd configuration file "+$dhcpd_file
      command = "touch #{$dhcpd_file}"
      execute_command(message,command)
    end
  end
  return
end

# Build DHCP plist file

def create_osx_dhcpd_plist()
  xml_output = []
  tmp_file   = "/tmp/plist.xml"
  plist_name = "homebrew.mxcl.isc-dhcp"
  plist_file = "/Library/LaunchDaemons/homebrew.mxcl.isc-dhcp.plist"
  dhcpd_bin  = "/usr/local/sbin/dhcpd"
  message    = "Checking:\tDHCPd configruation"
  command    = "cat #{plist_file} | grep '#{$default_net}'"
  output     = execute_command(message,command)
  if !output.match(/#{$default_net}/)
    xml = Builder::XmlMarkup.new(:target => xml_output, :indent => 2)
    xml.instruct! :xml, :version => "1.0", :encoding => "UTF-8"
    xml.declare! :DOCTYPE, :plist, :PUBLIC, :'"-//Apple Computer//DTD PLIST 1.0//EN"', :'"http://www.apple.com/DTDs/PropertyList-1.0.dtd"'
    xml.plist(:version => "1.0") {
      xml.dict {
        xml.key("label")
        xml.string(plist_name)
        xml.key("ProgramArguments")
        xml.array {
          xml.string(dhcpd_bin)
          xml.string($default_net)
          xml.string("-4")
          xml.string("-f")
        }
      }
      xml.key("Disabled") ; xml.false
      xml.key("KeepAlive") ; xml.true
      xml.key("RunAtLoad") ; xml.true
      xml.key("LowPriorityID") ; xml.true
    }
    file=File.open(tmp_file,"w")
    xml_output.each do |item|
      file.write(item)
    end
    file.close
    message = "Creating:\tService file "+plist_file
    command = "cp #{tmp_file} #{plist_file} ; rm #{tmp_file}"
    execute_command(message,command)
  end
  return
end

# Check ISC DHCP installed on OS X

def check_osx_dhcpd()
  check_osx_dhcpd_installed()
  create_osx_dhcpd_plist()
  service = "dhcp"
  check_osx_service_is_enabled(service)
  return
end

# Enable OS X service

def refresh_osx_service(service_name)
  if !service_name.match(/\./)
    if service_name.match(/dhcp/)
      service_name = "homebrew.mxcl.isc-"+service_name
    else
      service_name = "com.apple."+service_name+"d"
    end
  end
  disable_osx_service(service_name)
  enable_osx_service(service_name)
  return
end

# Enable OS X service

def enable_osx_service(service_name)
  check_osx_service_is_enabled(service_name)
  message = "Enabling:\tService "+service_name
  command = "launchctl start #{service_name}"
  output  = execute_command(message,command)
  return output
end

# Enable OS X service

def disable_osx_service(service_name)
  message = "Disabling:\tService "+service_name
  command = "launchctl stop #{service_name}"
  output  = execute_command(message,command)
  return output
end
